### DISCUSSION TITLE
Louvain algorithm

### DISCUSSION THREAD

**Discussion by marinaffc** [sknetwork-team/scikit-network#494] [Q&A] [answered]
Hi SkNetwork team, 

I am using your tool for a project to partition a graph based on modularity maximisation and I was wondering if it is possible to obtain intermediate steps for Louvain algorithm. 

If I have a graph, I would like to know what is the best modularity and clustering for 2 partitions, for 3 partitions, for 4 partitions...and so on. Is this possible at all?

Thank you in advance!

**Comment by QLutz** [comment, marked as answer]
Hello,

Getting the intermediate steps for Louvain is possible in a somewhat cumbersome way: you can use the `n_aggregations` parameter with increasing values to see the results after each aggregation phase. To get the aggregated graph, use `return_aggregate`.

By n-partitions, do you mean partitions with exactly/at most n clusters? If that is the case, this is one of the known problems of the Louvain method: you have to tweak the `resolution` parameter to change the number of clusters that are obtained (this number increases with the resolution). To set a precise number of clusters, you can either use the [k-means](https://scikit-network.readthedocs.io/en/latest/reference/clustering.html#k-means) algorithm (which is applied on an embedding of the nodes of the graph) or better yet, use one of the [hierarchical clustering](https://scikit-network.readthedocs.io/en/latest/reference/hierarchy.html#hierarchy) methods which enable you to fit your graph once and cut with the desired number of clusters. One downside in your case is that those algorithms do not directly optimise the modularity like Louvain.

Best,
